### PR TITLE
Composed Tasks

### PR DESCRIPTION
**Pull request by BoykoAlex**
Composed Tasks UI reanimated from Flo for Spring XD

Resolves spring-cloud/spring-cloud-dataflow-ui#170

### CODE REVIEW

File: ui/app/scripts/task/controllers/definitions.js [hunk 1]
```diff
@@ -90,6 +118,31 @@ define(['model/pageable'], function (Pageable) {
       $scope.bulkDefineTasks = function() {
         $state.go('home.tasks.bulkDefineTasks');
       };
+      $scope.isComposedTask = function(item) {
+        return item && angular.isString(item.dslText) && item.dslText.startsWith('composed-task-runner ');
```
Comment (ilayaperumalg) [inline ui/app/scripts/task/controllers/definitions.js]: The name `composed-task-runner` isn't a fixed name for all the composed task and one can define custom name for this via dataflow property. Hence, we need a better approach to decide if the task is composed or not.

Comment (ilayaperumalg) [inline ui/app/scripts/task/controllers/definitions.js]: May be we can look for `--graph` and probably some syntax that is applicable only for the composed task?

File: ui/app/scripts/app/controllers/apps.js [hunk 2]
```diff
@@ -81,7 +81,7 @@ define(['model/pageable'], function (Pageable) {
 
                                 // Define function for obtaining properties of a app
                                 item.getProperties = function() {
-                                    return appService.getAppInfo(this.type, this.name);
+                                    return appService.getAppInfo(this.type, this.name).$promise;
```
Comment (ilayaperumalg) [inline ui/app/scripts/app/controllers/apps.js]: While it definitely makes sense fix this to return Promise, I wonder what made this work fine so far? ;)

File: ui/app/scripts/app/services.js [hunk 3]
```diff
@@ -73,10 +74,15 @@ define(['angular', 'lodash'], function (angular, _) {
...
-                    return $http({
-                        method: 'GET',
-                        url: $rootScope.dataflowServerUrl + '/apps/'+appType+'/'+appName
-                    });
+                    return $resource($rootScope.dataflowServerUrl + '/apps/'+appType+'/'+appName, {}, {
+                        query: {
+                            method: 'GET'
+                        }
+                    }).get();
+                    // return $http({
```
Comment (ilayaperumalg) [inline ui/app/scripts/app/services.js]: Please remove the commented out lines (on this PR) unless they are really needed for references.

File: ui/app/scripts/routes.js [hunk 4]
```diff
@@ -306,6 +306,24 @@ define(['./app'], function (dashboard) {
...
+      url : 'tasks/definitions/{taskName}',
+      templateUrl : taskTemplatesPath + '/detailed-task.html',
+      controller: 'DetailedTaskController',
+      data:{
+        authenticate: true,
+        feature: 'tasksEnabled'
+      }
+    })
+    .state('home.tasks.tabs.createComposedTask', {
+      url: '/createComposedJob',
```
Comment (ilayaperumalg) [inline ui/app/scripts/routes.js]: Is this URL defined as a data flow server REST endpoint already?

File: ui/app/scripts/app/controllers/apps.js (hunk 2 above)
Comment (BoykoAlex) [inline ui/app/scripts/app/controllers/apps.js]: `$http` service returns a promise already. I have switched it to `$resource` service that seems more appropriate and it returns an object with a `$promise` property.

File: ui/app/scripts/routes.js (hunk 4 above)
Comment (BoykoAlex) [inline ui/app/scripts/routes.js]: Nope. The endpoint used to be `tasks/composed-definitions` but it's gone now. The regular task definition creation endpoint `tasks/definitions` creates composed task definition now.

File: ui/app/scripts/app/services.js (hunk 3 above)
Comment (BoykoAlex) [inline ui/app/scripts/app/services.js]: Sounds good.

File: ui/app/scripts/task/controllers/definitions.js (hunk 1 above)
Comment (BoykoAlex) [inline ui/app/scripts/task/controllers/definitions.js]: Thanks! Good point! I'll look for `--graph` at least... maybe something more...

**Comment by BoykoAlex** [comment]
@ilayaperumalg Updated the PR per your comments. Lts just look for `--graph="XXX"` in the DSL. If `graph` property is there it's fair to show the graph

**Comment by BoykoAlex** [comment]
Squashed commits

File: ui/app/scripts/app/controllers/apps.js (hunk 2 above)
Comment (ilayaperumalg) [inline ui/app/scripts/app/controllers/apps.js]: ok, I see that in `getAppInfo` service now. thanks!

File: ui/app/scripts/routes.js (hunk 4 above)
Comment (ilayaperumalg) [inline ui/app/scripts/routes.js]: ok, In that case, what does this `/createComposedJob` do?

Comment (BoykoAlex) [inline ui/app/scripts/routes.js]: It opens up an `Create Composed Task` editor UI tab. I guess it shouldn't be `/createComposedJob`. I have changed it to `/create-composed-task`. It's the same as `/definitions` URL for task definitions tab contents page. The full URL would have `/tasks/` prefix.

**Comment by trisberg** [comment]
When I create a stream I can type in a DSL like:

```
ticktock = time | log
```

Trying something similar with composed tasks:

```
foobar = foo && bar
```

but that's not accepted

Also, when I create a composed task I see the message that task has been created and the graph is erased, could we show the graph or switch to the definitions view instead of staying at a blank Create Composed Task view

**Comment by BoykoAlex** [comment]
@trisberg I think that is because DSL doesn't allow the name to be in the definition. User is allowed to enter the name of the task when creation dialog is shown. Although, I agree that perhaps we should have been consistent with streams I doubt this can be easily fixed now.
Navigating to `definitions` page after creation is debatable... It's not done for streams either. Also we'd have to navigate to the last page of definitions probably to show it... Not sure... In any case I'd rather push all these changes now and then polish and add small features on top of it.

**Comment by aclement** [comment]
I take Thomas' point about supporting the same as streams "foo = aaa && bbb", we should do that but I'm ok with that not being in MVP. Worth tackling when we get into the DSL syntax post this version because we have a lot to discuss there.

**Comment by ilayaperumalg** [comment]
LGTM,merging.

**Comment by ilayaperumalg** [comment]
Squashed, rebased and merged as fa84467beacdd90bfd855087d6baffd6c061248e

**Pull request closed without merging**